### ISSUE TITLE
tracing: client interceptor fails to replace parent span on the wire

### ISSUE THREAD

**Issue by tbg** [cockroachdb/cockroach#96173]
**Describe the problem**

I think this code

https://github.com/cockroachdb/cockroach/blob/5c38cfb6ab518ac2b8354c3a2a238ff3cfc88844/pkg/util/tracing/grpcinterceptor/grpc_interceptor.go#L228-L256

needs to wrap the `ctx` in scope with the newly created client span. Otherwise, the remote side will create a child span from the parent, in effect creating a sibling of the outgoing gRPC client interceptor rather than a child. This is confusing, as it leads to a picture like this:

<img width="1310" alt="image" src="https://user-images.githubusercontent.com/5076964/215480139-bc09db66-59c8-4351-bf5d-d8f878ee7710.png">

where the second span should really be a child of the first. This caused a good amount of time wasted in https://github.com/cockroachlabs/support/issues/2035.

NB: I observed this problem on 22.1.10 but the code snippet above is 23.1 (not sure when it was introduced). The code is quite different but the problem seems to exist still.

Jira issue: CRDB-23982

Epic CRDB-20779

**Comment by tbg** [comment]
Btw there's a chance I'm somehow missing how this works. In https://github.com/cockroachlabs/support/issues/2035#issuecomment-1408668863 I am confused why this client span discussed here can outlive the caller's (DistSender) span. IMO this ought to be impossible, except maybe in context cancellation scenarios, of which this is not one. (The query ultimately completes successfully).

**Comment by andreimatei** [comment]
You're right that for `Batch` RPCs the client span and the server span end up being siblings, which is not great.
However that's not exactly the fault of the `ClientInterceptor` that you point to. As far as this interceptor is concerned, when the trace information is propagated to the server, it propagates the correct child span, [here](https://github.com/cockroachdb/cockroach/blob/10ef5d9d7a646831799faa0f0611834fe032b93b/pkg/util/tracing/grpcinterceptor/grpc_interceptor.go#L244). So most RPCs get the right info on the server side.
The thing is that `Batch` and `SetupFlow` and a few other RPCs are special, and they don't rely on this interceptor to use gRPC metadata (which is expensive). Instead, the metadata is inhibited through `methodExcludedFromTracing()`, and the trace info is included explicitly in dedicated fields on the requests. For `Batch`, this is done [here](https://github.com/cockroachdb/cockroach/blob/10ef5d9d7a646831799faa0f0611834fe032b93b/pkg/rpc/nodedialer/nodedialer.go#L317).
There's also the importing of the remote span part, which is only implemented for `Batch`. That's done [here](https://github.com/cockroachdb/cockroach/blob/53f394da85573530c38c4197be2dacc6d42af10e/pkg/kv/kvclient/kvcoord/transport.go#L236).
Both of these are done above the interceptor, so they don't have access to the client span created by the interceptor. 

Things are done this way because there's no completely obvious way for the interceptor to do the setting of the parent ID and then importing of the response span in a generic way. You can imagine the requests that support tracing implementing some interface - but one of the things this interface would need to support is a way to make a shallow copy of the request (because of the race transport, which is not necessarily a very good reason). Some requests need this copy, other don't. So I hesitated to put this bespoke stuff in the interceptor.
Also I for one was never motivated to care about whether the server span is a child of the client. The client span is rarely interesting (it doesn't have any logs), so I don't necessarily want to click through it every time I want to get to more interesting data. The only reason why it even exists is to show cases exactly like in your investigation - when the server's response is processed with high latency by the client.

WDYT?

**Comment by tbg** [comment]
Thanks for the explanation. I agree, in 99.99% of cases we don't get any value from the client span and it would be annoying to have to expand through it. But it just cost us probably like 4 eng hours because Irfan and I we didn't realize this was the client span... 🙈 so would be good to do something.

Could we change this client interceptor to not create a span, but instead put an event in the existing span at interception and post interception time? As you mention the child never has any interesting events; its raison d'etre is just the duration measurement. I think this would check all of the boxes; we don't get confused and we don't make useless spans and still we get to see when something in the grpc client code causes a delay.

**Comment by andreimatei** [comment]
I don't think turning the span into events is a great idea. When the parent span does many RPCs, the events in the parent won't help too much. If I need to do something, I think I'd rather make the server a child span.
There are already clues indicating that one span is the client and another is the server - the client is on the same node as its parent, and it also has a client tag. But we could also put it in the name if you want.

**Comment by tbg** [comment]
Ok then let's make the name of the span very obvious - I think @irfansharif and I wasted time because the only indication was the `span.kind` tag, which isn't very obvious. Maybe we can also link from it to this issue somehow? Maybe naming it `/Foo/Bar-see-issue-96173 or something like that, but maybe that's too obnoxious.
